### PR TITLE
chore: upgrade react 18

### PR DESCRIPTION
**Pull request by daibhin**
## Problem

Upgrade to React 18. Moving on from https://github.com/PostHog/posthog/pull/16605

## Changes

Relatively little

## How did you test this code?

Clicked around

Screenshot 2023-10-03 at 11 36 20

Want to see what the tests do 👀

### CODE REVIEW

**Comment by daibhin** [comment]
Still working on this. Might get around to it after notebooks launch

**Comment by daibhin** [comment]
@PostHog/noteforce-3000 @pauldambra just tagged you both for a review because it would be good to get some eyes on this now that tests & checks are passing

File: frontend/__snapshots__/posthog-3000-sidebar--dashboards.png
Comment (pauldambra) [inline frontend/__snapshots__/posthog-3000-sidebar--dashboards.png:1]: related to this change or broken in master?

File: frontend/__snapshots__/scenes-app-experiments--running-trend-experiment.png
Comment (pauldambra) [inline frontend/__snapshots__/scenes-app-experiments--running-trend-experiment.png:1]: is the overlapping text on the participant pill introduced in this PR

File: frontend/src/lib/lemon-ui/LemonButton/LemonButton.tsx [hunk 1]
```diff
@@ -26,7 +26,7 @@ export interface LemonButtonPropsBase
         | 'role'
         | 'aria-haspopup'
     > {
-    children?: React.ReactNode
+    children?: React.ReactNode | undefined
```
Comment (pauldambra) [inline frontend/src/lib/lemon-ui/LemonButton/LemonButton.tsx:29]: weird why do we need `blah?: Type | undefined` on this line and not the ones below? I thought the `?` _was_implicitly `| undefined`

File: frontend/src/scenes/insights/filters/ActionFilter/RenameModal.tsx [hunk 2]
```diff
@@ -6,6 +6,7 @@ import { Button, Input, Modal } from 'antd'
 import { getDisplayNameFromEntityFilter } from 'scenes/insights/utils'
 import { renameModalLogic } from 'scenes/insights/filters/ActionFilter/renameModalLogic'
 import { InputFocusOptions } from 'antd/lib/input/Input'
+import InputRef from 'antd/lib/input/Input'
```
Comment (pauldambra) [inline frontend/src/scenes/insights/filters/ActionFilter/RenameModal.tsx:9]: Is it better to replace the antd input instead?

**Review by pauldambra — APPROVED** [review]
clicking around the UI seems to work, no obvious new errors in the log

🚀

File: frontend/__snapshots__/scenes-app-experiments--running-trend-experiment.png
Comment (daibhin) [reply frontend/__snapshots__/scenes-app-experiments--running-trend-experiment.png:1]: Confirmed this is the case on `master` too. Will follow up with a proper fix when I remove all the `Row` and `Col` components from antd
Screenshot 2023-10-27 at 12 04 02

File: frontend/__snapshots__/posthog-3000-sidebar--dashboards.png
Comment (daibhin) [reply frontend/__snapshots__/posthog-3000-sidebar--dashboards.png:1]: Looks to happen on `master` too

Screenshot 2023-10-27 at 12 07 01

Less sure about the cause of this one, maybe @Twixes has some thoughts. It seems to only happen if you navigate to that scene in the Storybook sidebar. Reloading the page and everything works as expected

Comment (thmsobrmlr) [reply frontend/__snapshots__/posthog-3000-sidebar--dashboards.png:1]: Mmh this might be related to the dark mode toggle the. I'm forcing the feature flag in the story, but maybe that doesn't work when something else was already loaded.

Comment (thmsobrmlr) [reply frontend/__snapshots__/posthog-3000-sidebar--dashboards.png:1]: Seems to be something else, but an easy fix/workaround https://github.com/PostHog/posthog/pull/18243

File: frontend/src/lib/lemon-ui/LemonButton/LemonButton.tsx (hunk 1 above)
Comment (daibhin) [reply frontend/src/lib/lemon-ui/LemonButton/LemonButton.tsx:29]: Yeah, this actually came up in a different PR with type changes I was working on. It's not needed so removing

File: frontend/src/scenes/insights/filters/ActionFilter/RenameModal.tsx (hunk 2 above)
Comment (daibhin) [reply frontend/src/scenes/insights/filters/ActionFilter/RenameModal.tsx:9]: Yep, done in https://github.com/PostHog/posthog/pull/18242

File: frontend/__snapshots__/scenes-app-insights--trends-line-edit.png
Comment (pauldambra) [inline frontend/__snapshots__/scenes-app-insights--trends-line-edit.png:1]: is the change here where the filter boxes are now full width flapping or introduced in this PR?

Comment (daibhin) [reply frontend/__snapshots__/scenes-app-insights--trends-line-edit.png:1]: Looks like it's reverted in the latest run so I'm guessing it's just a little flappy

File: package.json [hunk 3]
```diff
@@ -221,14 +218,13 @@
         "@types/pixelmatch": "^5.2.4",
         "@types/pngjs": "^6.0.1",
         "@types/query-selector-shadow-dom": "^1.0.0",
-        "@types/react": "^16.14.2",
-        "@types/react-dom": "^16.9.8",
+        "@types/react": "^17.0.39",
+        "@types/react-dom": "^18.2.8",
```
Comment (Twixes) [inline package.json:222]: These should have `^18.2.0` just like their respective packages

File: frontend/src/toolbar/button/Circle.tsx [hunk 4]
```diff
@@ -7,7 +7,7 @@ interface CircleProps extends React.PropsWithoutRef<JSX.IntrinsicElements['div']
     extensionPercentage?: number
     distance?: number
     rotation?: number
-    content?: string | JSX.Element
```
Comment (Twixes) [inline frontend/src/toolbar/button/Circle.tsx:10]: Why this change?

File: frontend/src/scenes/funnels/useFunnelTooltip.tsx [hunk 5]
```diff
@@ -116,7 +116,8 @@ export function useFunnelTooltip(showPersonsModal: boolean): React.RefObject<HTM
         }
         const tooltipRect = tooltipEl.getBoundingClientRect()
         if (tooltipOrigin) {
-            ReactDOM.render(
+            // TODO: Shouldn't we unmount at some point??
+            createRoot(tooltipEl).render(
```
Comment (Twixes) [inline frontend/src/scenes/funnels/useFunnelTooltip.tsx:120]: This legacy architecture of the insight tooltip is horrible, and React notices… any time the tooltip rendered, this shows up in the console:
`Warning: You are calling ReactDOMClient.createRoot() on a container that has already been passed to createRoot() before.`
In the world map it's easy to rack up a thousand of these warnings, so that's probably the last thing we should solve before shipping this upgrade

File: package.json (hunk 3 above)
Comment (daibhin) [reply package.json:222]: It originally was but caused a ton of type issues, caused by this issue I believe: https://github.com/facebook/react/issues/24304.

As far as I could tell a lot of our antd components were affected. I tried upgrading antd to confirm but that was a task in and of itself.

Is there another way around this? Or any reason it can't stay at v17 temporarily

File: frontend/src/toolbar/button/Circle.tsx (hunk 4 above)
Comment (daibhin) [reply frontend/src/toolbar/button/Circle.tsx:10]: It was throwing an error previously because the `JSX.IntrinsicElements['div']` type from which it extends also has a `content` prop. Just reverted it and eslint seems happy so will try without it again.

File: frontend/src/scenes/funnels/useFunnelTooltip.tsx (hunk 5 above)
Comment (daibhin) [reply frontend/src/scenes/funnels/useFunnelTooltip.tsx]: Would something like this work? https://github.com/PostHog/posthog/pull/17731/commits/69ed960277cf149cdc09c793a88f536b3892d583

Tooltips seem to render fine and no console errors 🎆 
Oct-31-2023 18-51-57

**Review by Twixes — APPROVED** [review]
This looks good!
There's a worrying amount of UI snapshot flip-flopping here, and we need to fix that. However, I'm not sure if that's related to this PR

**Review by benjackwhite — APPROVED** [review]
Clicked around a bunch and this feels great. I say lets go for it
